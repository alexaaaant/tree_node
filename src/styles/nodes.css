.nodes_container {
    display: flex; 
    justify-content: center;  
}

.node {
    margin: 5px;
}

.node__name {
    cursor: pointer;
    border: 1px solid gray;;
}

.node__name_selected {
    border-color: red;
}

.node_children {
    margin-left: 10px;
}