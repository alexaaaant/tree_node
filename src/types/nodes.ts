export interface NodeRaw {
  id: number;
  parent?: number;
  name: string;
}

export interface Node extends NodeRaw {
  children?: Array<NodeRaw['id']>;
  selected?: boolean;
}

export interface NodeState {
  rootNodes: Array<Node['id']>;
  nodes: Record<Node['id'], Node>;
}
