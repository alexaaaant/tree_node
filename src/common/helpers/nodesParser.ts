import { Node, NodeRaw, NodeState } from '../../types/nodes';

export const rebuildTree = (rawNodes: NodeRaw[]): NodeState => {
    const dict: Record<NodeRaw['id'], Node> = {};
    const rootNodes: Array<Node['id']> = [];

    rawNodes.forEach((rawNode) => {
        dict[rawNode.id] = { ...rawNode, selected: false };

        if (rawNode.parent !== undefined) {
            const parentNode = dict[rawNode.parent];
            
            const children = parentNode.children || [];
            children.push(rawNode.id);

            parentNode.children = children;
        } else {
            rootNodes.push(rawNode.id);
        }
    });
    return { rootNodes, nodes: dict };
};
