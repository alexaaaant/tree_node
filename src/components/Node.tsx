import { FC, memo } from 'react';
import { useDispatch } from '../common/hooks/useDispatch';
import { useSelector } from '../common/hooks/useSelector';
import { groupSelected, nodeSelected } from '../slices/nodesSlice';
import { Node } from '../types/nodes';

interface NodeProps {
  nodeId: Node['id'];
}

export const NodeComponent: FC<NodeProps> = memo(({ nodeId }) => {
    const node = useSelector((state) => state.nodes.nodes[nodeId], 
                             (prevNode, newNode) => prevNode.selected === newNode.selected);
    const dispatch = useDispatch();

    const selectNode = () => {
      dispatch(nodeSelected(nodeId));
    };

    const selectGroup = (event: React.MouseEvent<HTMLDivElement>) => {
      event.preventDefault(); 
      dispatch(groupSelected(nodeId));
    };

    return (
      <div className="node">
        <div
          className={`node__name ${node.selected ? 'node__name_selected' : ''}`}
          onClick={selectNode}
          onContextMenu={selectGroup}
        >
          {node.name}
        </div>
        {node.children ? (
          <div className="node_children">
            {node.children.map((childId) => (
              <NodeComponent key={childId} nodeId={childId} />
            ))}
          </div>
        ) : (
          <></>
        )}
      </div>
    );
  });

NodeComponent.displayName = 'NodeComponent';
