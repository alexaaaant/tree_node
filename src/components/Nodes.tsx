import { NodeComponent } from './Node';
import { useEffect } from 'react';
import { rebuildTree } from '../common/helpers/nodesParser';
import rawNodes from '../nodes';
import { useDispatch } from '../common/hooks/useDispatch';
import { nodesAdded } from '../slices/nodesSlice';
import { useSelector } from '../common/hooks/useSelector';
import '../styles/nodes.css';

export const NodesContainer = () => {
  const dispatch = useDispatch();
  const rootNodes = useSelector((state) => state.nodes.rootNodes);

  useEffect(() => {
    const nodesData = rebuildTree(rawNodes);
    dispatch(nodesAdded(nodesData));
  },        []);

  return (
    <div className="nodes_container">
      {rootNodes.map((nodeId) => (
        <NodeComponent key={nodeId} nodeId={nodeId} />
      ))}
    </div>
  );
};
