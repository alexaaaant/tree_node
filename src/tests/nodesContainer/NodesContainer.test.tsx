import { NodesContainer } from '../../components/Nodes';
import { NodeRaw } from '../../types/nodes';
import { fireEvent, render, screen } from '../helpers/test-utils';

jest.mock('../../nodes', () => ({
  __esModule: true,
  default: [
    { id: 0, name: '0' },
    { id: 1, name: '1' },
    { id: 2, name: '2', parent: 0 },
    { id: 3, name: '3', parent: 0 },
    { id: 4, name: '4', parent: 1 },
    { id: 5, name: '5', parent: 3 },
  ] as NodeRaw[],
}));

describe('NodesContainer', () => {
  it('should render a full tree, if a nodesData exists', () => {
    const { container } = render(<NodesContainer />);
    expect(container).toMatchSnapshot();
  });

  it('should select one node, if user clicks on node name', () => {
    const { container } = render(<NodesContainer />);

    const nodeElement = screen.getByText('3');
    expect(nodeElement).not.toHaveClass('node__name_selected');

    fireEvent.click(nodeElement);
    expect(nodeElement).toHaveClass('node__name_selected');

    // eslint-disable-next-line testing-library/no-container
    const selectedElements = container.querySelectorAll('.node__name_selected');
    expect(selectedElements).toHaveLength(1);
  });

  it('should select all group, if user click by right button on node name', () => {
    const { container } = render(<NodesContainer />);

    const nodeElement = screen.getByText('0');
    
    fireEvent.contextMenu(nodeElement);
    expect(nodeElement).toHaveClass('node__name_selected');

    // eslint-disable-next-line testing-library/no-container
    const selectedElements = container.querySelectorAll('.node__name_selected');
    expect(selectedElements).toHaveLength(3);

    expect(container).toMatchSnapshot();
  });
});
