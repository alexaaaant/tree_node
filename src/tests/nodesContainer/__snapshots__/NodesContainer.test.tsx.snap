// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodesContainer should render a full tree, if a nodesData exists 1`] = `
<div>
  <div
    class="nodes_container"
  >
    <div
      class="node"
    >
      <div
        class="node__name "
      >
        0
      </div>
      <div
        class="node_children"
      >
        <div
          class="node"
        >
          <div
            class="node__name "
          >
            2
          </div>
        </div>
        <div
          class="node"
        >
          <div
            class="node__name "
          >
            3
          </div>
          <div
            class="node_children"
          >
            <div
              class="node"
            >
              <div
                class="node__name "
              >
                5
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="node"
    >
      <div
        class="node__name "
      >
        1
      </div>
      <div
        class="node_children"
      >
        <div
          class="node"
        >
          <div
            class="node__name "
          >
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NodesContainer should select all group, if user click by right button on node name 1`] = `
<div>
  <div
    class="nodes_container"
  >
    <div
      class="node"
    >
      <div
        class="node__name node__name_selected"
      >
        0
      </div>
      <div
        class="node_children"
      >
        <div
          class="node"
        >
          <div
            class="node__name node__name_selected"
          >
            2
          </div>
        </div>
        <div
          class="node"
        >
          <div
            class="node__name node__name_selected"
          >
            3
          </div>
          <div
            class="node_children"
          >
            <div
              class="node"
            >
              <div
                class="node__name "
              >
                5
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="node"
    >
      <div
        class="node__name "
      >
        1
      </div>
      <div
        class="node_children"
      >
        <div
          class="node"
        >
          <div
            class="node__name "
          >
            4
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
